File: C++/代码随想录/Dynamic_Programming/Subsequence_problem/647. Palindromic Substrings.py
def countSubstrings(s):
    n = len(s)
    dp = [[False] * n for _ in range(n)]
    count = 0

    # 单个字符总是回文串
    for i in range(n):
        dp[i][i] = True
        count += 1

    # 检查长度为 2 的子串
    for i in range(n - 1):
        if s[i] == s[i + 1]:
            dp[i][i + 1] = True
            count += 1

    # 检查长度大于 2 的子串
    for length in range(3, n + 1):
        for i in range(n - length + 1):
            j = i + length - 1
            if s[i] == s[j] and dp[i + 1][j - 1]:
                dp[i][j] = True
                count += 1

    return count

# 测试用例
print(countSubstrings("abc"))  # 输出应该是 3
print(countSubstrings("aaa"))  # 输出应该是 6

File: C++/代码随想录/Dynamic_Programming/Subsequence_problem/647. Palindromic Substrings.cpp
// 647. Palindromic Substrings
// https://leetcode.com/problems/palindromic-substrings/

#include <iostream>
#include <string>
#include <vector>

int countSubstrings(std::string s) {
    int n = s.length();
    std::vector<std::vector<bool>> dp(n, std::vector<bool>(n, false));
    int count = 0;

    for (int i = n - 1; i >= 0; i--) {
        for (int j = i; j < n; j++) {
            if (s[i] == s[j] && (j - i < 2 || dp[i + 1][j - 1])) {
                dp[i][j] = true;
                count++;
            }
        }
    }

    return count;
}

int main() {
    std::vector<std::string> testCases = {
        "abc",
        "aaa",
    };

    for (int i = 0; i < testCases.size(); i++) {
        std::string s = testCases[i];

        int count = countSubstrings(s);

        std::cout << "Number of palindromic substrings for test case " << (i + 1) << ": ";
        std::cout << count << std::endl;
    }

    return 0;
}